drivers: dac: API + STM32L0 driver implementation

martinjaeger opened the pull request:
Fixes #3076. 

Here is a suggestion to implement the DAC driver.

As it is the first time I'm implementing a driver for Zephyr, please be forgiving if something was not done the usual or correct way. Just let me know what can be improved.

I tested the driver with an STM32L073RZ Nucleo board. As soon as we agree on the API, i can also add additional STM32 MCUs. I've got a board with STM32G431 for testing aswell.

Unfortunately, I don't know very much about DAC features of other manufacturer's MCUs. I decided not to include special STM features like sawtooth wave generation into the generic API (not sure if that's a typical feature). Also DMA is currently not supported. The only configuration is whether a channel should be buffered or not.

Similar to the ADC driver, the resolution is defined for each write request rather than once per channel at the beginning.

martinjaeger committed:
boards: nucleo_l073rz: Add DAC peripheral driver

Adds support for the new DAC driver to ST Nucleo board with
STM32L073RZ MCU.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
samples: drivers: dac: Add API example

This example generates a sawtooth signal to show how to use the DAC.

Signed-off-by: Martin Jäger <martin@libre.solar>

erwango reviewed (changes requested):
Thanks for this welcome addition. Seems like a pretty good start.
Can you add one more commit with the api only ?

martinjaeger committed:
drivers: dac: Add API for DAC peripherals

DAC (digital to analog converter) peripheral driver with a generic API
suitable for most MCUs (only basic DAC features considered).

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
drivers: dac: Add support for STM32L0 series

First implementation for STM32L0 series MCUs to be used for testing.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented:
> Can you add one more commit with the api only ?
Ok, makes sense. Just split the first commit.

Above checkpatch warning "TRAILING_SEMICOLON: macros should not use a trailing semicolon" is actually a false positive, as it will not compile with the trailing semicolon removed. Maybe the parser can't handle such complex defines? To whom should this be reported?

jfischer-phytec-iot commented:
> > Can you add one more commit with the api only ?
> 
> Ok, makes sense. Just split the first commit.
> 
> Above checkpatch warning "TRAILING_SEMICOLON: macros should not use a trailing semicolon" is actually a false positive, as it will not compile with the trailing semicolon removed. Maybe the parser can't handle such complex defines? To whom should this be reported?

Why is it false positive?

```diff
diff --git a/drivers/dac/dac_stm32.c b/drivers/dac/dac_stm32.c
index 42a3727c8a..c01b090818 100644
--- a/drivers/dac/dac_stm32.c
+++ b/drivers/dac/dac_stm32.c
@@ -134,21 +134,21 @@ static struct dac_stm32_data dac_stm32_data_##index = {           
        \
 DEVICE_AND_API_INIT(dac_##index, DT_DAC_##index##_NAME, &dac_stm32_init,\
                    &dac_stm32_data_##index, &dac_stm32_cfg_##index,    \
                    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,   \
-                   &api_stm32_driver_api);                             \
+                   &api_stm32_driver_api)
 
 
 #ifdef CONFIG_DAC_1
-STM32_DAC_INIT(1)
+STM32_DAC_INIT(1);
 #endif /* CONFIG_DAC_1 */
 
 #ifdef CONFIG_DAC_2
-STM32_DAC_INIT(2)
+STM32_DAC_INIT(2);
 #endif /* CONFIG_DAC_2 */
 
 #ifdef CONFIG_DAC_3
-STM32_DAC_INIT(3)
+STM32_DAC_INIT(3);
 #endif /* CONFIG_DAC_3 */
 
 #ifdef CONFIG_DAC_4
-STM32_DAC_INIT(4)
+STM32_DAC_INIT(4);
 #endif /* CONFIG_DAC_4 */

```

btw, none of DAC are enabled in the sample.

martinjaeger commented:
> Why is it false positive?

Cool, didn't realize that possible fix. I took parts of this from [ADC driver](https://github.com/zephyrproject-rtos/zephyr/blob/0c3fbaa8822f63eab9cd1ad97e82ee726fd63348/drivers/adc/adc_stm32.c#L664), but it has a `}` at the end of the define, so that's why checkpatch doesn't complain there.

Will wait for further comments and fix this in the next force-push.

> btw, none of DAC are enabled in the sample.

Oops, you are right. I had `DAC_1` configured as `default y` in `drivers/dac/Kconfig.stm32` before, but removed it and didn't realize it's not enabled in the sample anymore. My .config had it enabled already.

Where should it be enabled ideally? In the sample's `proj.conf` or in the board `Kconfig.defconfig` if `DAC` is enabled?

jfischer-phytec-iot commented:
> Where should it be enabled ideally? In the sample's `proj.conf` or in the board `Kconfig.defconfig` if `DAC` is enabled?

I think in the board `Kconfig.defconfig` if `DAC` is enabled. :+1:

erwango reviewed (changes requested):
Few points otherwise LGTM.
Would be nice to have inputs from potential users.

martinjaeger commented on soc/arm/st_stm32/stm32l0/dts_fixup.h:
```diff
@@ -149,4 +149,9 @@
+#define DT_DAC_1_NAME			DT_ST_STM32_DAC_40007400_LABEL
+#define DT_DAC_1_CLOCK_BITS		DT_ST_STM32_DAC_40007400_CLOCK_BITS_0
+#define DT_DAC_1_CLOCK_BUS		DT_ST_STM32_DAC_40007400_CLOCK_BUS_0
+
```
The STM32L0 has only 1 DAC. Will add further instances in other STM32 series if applicable.

martinjaeger commented on drivers/dac/dac_stm32.c:
```diff
@@ -0,0 +1,154 @@
+	 * No independent internal reference available for DAC peripheral,
+	 * DAC uses voltage at Vref+ that may be set by external hardware.
+	 */
+	.ref_internal = 0
```
Yes, that would probably make sense. In this case it could be overwritten more easily depending on board specification. I'll have a look into this.

martinjaeger commented on samples/drivers/dac/src/main.c:
```diff
@@ -0,0 +1,40 @@
+	}
+
+	/* set half of supply/reference voltage */
+	dac_write(dac_dev, DAC_1ST_CHANNEL_ID, 2048, 12);
```
Yep, no problem. What about generating a saw-tooth signal in a for loop with an interval of one minute so that you can still observe it with a multi-meter?

erwango replied:
Indeed, maybe something a 1 sec period would be ok. It can even be seen using a led.

We can also think of something bit more elaborated and feedback the output to a gpio and verify input level. This way it could be tested with only a simple wire (could be left for future work).

martinjaeger replied:
1 sec period was too slow to be seen with a multimeter, so I took 4 sec (counting from 0 to 4095 with 1 msec sleep).

martinjaeger commented:
So if everyone is fine with the current status, I could add support for some other STM32 MCUs before we merge.

erwango commented:
> So if everyone is fine with the current status, I could add support for some other STM32 MCUs before we merge.

@martinjaeger It will be easier to merge in current state with fewer commits. Let's get this one merged first.

jfischer-phytec-iot commented:
> So if everyone is fine with the current status, I could add support for some other STM32 MCUs before we merge.

I think just STM32 is not enough, there should be at least one driver for an other platform.

erwango commented on drivers/dac/dac_stm32.c:
```diff
@@ -0,0 +1,155 @@
+#define CHAN(n)		LL_DAC_CHANNEL_##n
+static const u32_t table_channels[] = {
+	CHAN(1),
+	CHAN(2),
```
Please define under `#ifdef DAC_CHANNEL2_SUPPORT` or dts exacted flag.

martinjaeger replied:
Found that ll_dac.h defines LL_DAC_CHANNEL_2 only if it actually exists in hardware, so I'll use that as a flag.

In general, I'm not entirely sure how DTS and Kconfig should be configured to reflect actual features of a specific MCU. I see that peripherals appear in menuconfig even if the MCU doesn't support it (e.g. CAN in case of STM32L0). They are not selected by default, but what happens if a user enables it?
In our case, if someone enables DAC_3 even though it doesn't exist in hardware, we get a compile error because the defines are not found. But shouldn't we better prevent selecting an impossible configuration at all in menuconfig?

erwango replied:
> I see that peripherals appear in menuconfig even if the MCU doesn't support it (e.g. CAN in case of STM32L0). They are not selected by default, but what happens if a user enables it?

This is indeed an issue, ifnot supported by SoC, it should not be visible. It shouldbe fixed.

> But shouldn't we better prevent selecting an impossible configuration at all in menuconfig?

DTS will provide a more accurate way to declare this, but there are still few hickups to get it fully supported

martinjaeger replied:
Ok, so this seems to be a more general issue then if it affects many other peripheral drivers. Can we keep it as it is (i.e. similar to other drivers) here and discuss visibility of peripherals in DTS in general in a different issue?

erwango replied:
Sure, this is fine as is.

erwango commented on samples/drivers/dac/sample.yaml:
```diff
@@ -0,0 +1,11 @@
+    harness_config:
+      type: one_line
+      regex:
+        - "Counter received: (.*)"
```
Does not macth sample output:
```
DAC internal reference voltage: 3300 mV                                         
Generating sawtooth signal at DAC channel 1.  
```
Btw, if using multi_line, also add `ordered: true`

erwango commented on samples/drivers/dac/README.rst:
```diff
@@ -0,0 +1,32 @@
+
+The DAC output is defined in the board's devicetree and pinmux file.
+
+The first supported board is the nucleo_l073rz. More boards will follow.
```
Please reword to reflect current status.

martinjaeger replied:
What do you mean exactly?

erwango replied:
"Only board nucleo_l073rz is supported for now"

martinjaeger replied:
Ok, will update that with next force-push.

erwango commented on dts/arm/st/l0/stm32l0.dtsi:
```diff
@@ -180,6 +180,15 @@
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x20000000>;
+			status = "disabled";
+			label = "DAC_1";
+			#io-channel-cells = <1>;
```
Have you checked https://www.kernel.org/doc/Documentation/devicetree/bindings/iio/dac/st%2Cstm32-dac.txt ?
Ideally we should match this description.

martinjaeger replied:
I have checked it, but not sure if we can make it compatible. As far as I understand, we don't have a phandle for vref available, for example. Also the Zephyr ADC driver seems quite different from the Linux one. I took the ADC as a basis.

erwango reviewed (changes requested):
Additional comments after deeper review. These mainly targets stm32 implementation. 
It would be nice to have more reviews of potential users. Would you mind sending a mail to `zephyr-devel` mailing list to seek for more reviews?

Last, please add a doc/reference/peripherals/dac.rst to introduce API in doc.

martinjaeger commented:
Thanks @erwango for the detailed review. Will get back to this PR till end of the week, as I need to finish some other work first.

martinjaeger commented:
It seems like there are not too many people interested in the DAC driver. I asked for further review on the devel mailing list.

Do you still see a chance to get it merged for Zephyr 2.2 release? What else would be needed for that?

And should I really add my name under codeowners for the documentation .rst file or will we get someone responsible for documentation in general again?

erwango commented on samples/drivers/dac/sample.yaml:
```diff
@@ -0,0 +1,13 @@
+      ordered: true
+      regex:
+        - "DAC internal reference voltage: ([0-9]*) mV"
+        - "Generating sawtooth signal at DAC channel ([0-9]*).
```
Miss ' " ' at the end of the line

erwango reviewed (changes requested):
@martinjaeger, aside from the sanitycheck issues, I'm fine with the change.
So please fix the CI stuff and I'll approve.

martinjaeger commented:
> So please fix the CI stuff and I'll approve.

Done, except for the doc file in CODEOWNERS. I'm really not an expert in .rst and it was mentioned in Slack channel that it's not mandatory.

nashif commented:
rebase please

martinjaeger commented:
> rebase please

Done.

henrikbrixandersen commented on drivers/dac/Kconfig.stm32:
```diff
@@ -0,0 +1,35 @@
+	bool "STM32 DAC 4"
+	help
+	  Enable DAC 4. Each STM32 DAC has 2 channels.
+
```
It would be nice if we could just enable the instances based on their `status` property in the device tree.

martinjaeger replied:
Seems like a good idea. How can I access DTS properties in Kconfig? Is there an example from other drivers? Quickly looked at some other drivers, but didn't find one that implements this check.

erwango replied:
@henrikbrixandersen, the DT_INST_ solution is not in use in all STM32 peripheral drivers (and I'm personally not in favor of this solution). I'm awaiting https://github.com/zephyrproject-rtos/zephyr/issues/19285 to provide a sustainable solution to the peripheral instance activation issue. 
So this method is fine for now.

Also, CONFIG_DAC_X are needed anyway for pinmux definition.

henrikbrixandersen replied:
@martinjaeger Try grepping for `DT_INST_` under `drivers`. Many drivers use this approach.

henrikbrixandersen commented on dts/bindings/dac/dac-controller.yaml:
```diff
@@ -0,0 +1,19 @@
+    ref-internal:
+      type: int
+      required: true
+      description: Internal reference voltage in mV or 0 if not existing
```
Instead of `0 if not existing` I suggest changing it to `required: false`.

martinjaeger replied:
The internal reference is currently also exposed in the API. So if it's not set here, what should the API assume?
I think it makes sense to require defining it in the DTS. At least in the board-level DTS you should know if the Vref pin is wired to VCC or an external or internal reference.
In the Linux DTS they have a reference obtained from DTS in some way. But I'm not sure if that's possible in Zephyr aswell.

henrikbrixandersen replied:
How do you propose handling a DAC without any internal reference with the proposed API and DTS binding?

martinjaeger replied:
Agree that didn't make much sense to specify a fixed value in the DTS, so I removed ref-internal setting from DTS again. Actual value of the reference should be handled by the application.

henrikbrixandersen commented on dts/bindings/dac/st,stm32-dac.yaml:
```diff
@@ -0,0 +1,21 @@
+      const: 1
+
+io-channel-cells:
+    - input
```
Did you mean to write `output` here?

martinjaeger replied:
Yes, that makes more sense. Thanks.

henrikbrixandersen commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,144 @@
+struct dac_driver_api {
+	dac_api_channel_setup channel_setup;
+	dac_api_write         write;
+	u16_t ref_internal;	/* mV */
```
I believe this is too simple for a generic API. Something along the lines the ADC API methods for selecting a reference voltage would be ideal. The NXP KE1xF, for instance, supports two external DAC voltage references.

martinjaeger replied:
Which part of the API should we take from ADC API? I think the reference is as simple in ADC API as here.
Unfortunately, in STM32 the ADC internal reference is shared with DAC reference, so it's not possible to just set the reference voltage in the DAC driver. We might need something like a dedicated reference voltage driver then.

henrikbrixandersen replied:
The part where the user can select a reference: https://github.com/zephyrproject-rtos/zephyr/blob/1421dfb3f13602189210b590946433ec093c47cd/include/drivers/adc.h#L68

martinjaeger replied:
Made the reference handling more similar to the ADC now. For the mentioned NXP MCUs we can now select external 0 or 1. For STM32 I added `DAC_REF_DEFAULT`, as I think that `ADC_REF_INTERNAL` as used for the STM32 ADC is a bit misleading. The reference is in most cases not internal, but defined externally (either via VREF+ pin or VDDA). There is no possibility to select a dedicated internal ADC or DAC reference in STM32. It's only possible to drive the VREF+ pin by a buffer for some MCUs (e.g. G4).

erwango commented:
@martinjaeger, end of merge window for V2.2 is tomorrow.
Do you think you could do the requested modifications by tomorrow so e can try to get this change in ?

martinjaeger commented:
> @martinjaeger, end of merge window for V2.2 is tomorrow.
> Do you think you could do the requested modifications by tomorrow so e can try to get this change in ?

Oops, I thought it is end of next week.

Updates are done now, but we'll need another review of @henrikbrixandersen.

galak commented on soc/arm/st_stm32/stm32l0/dts_fixup.h:
```diff
@@ -149,4 +149,10 @@
 #define DT_ADC_1_CLOCK_BUS		DT_ST_STM32_ADC_40012400_CLOCK_BUS_0
 
+#define DT_DAC_1_BASE_ADDRESS		DT_ST_STM32_DAC_40007400_BASE_ADDRESS
+#define DT_DAC_1_NAME			DT_ST_STM32_DAC_40007400_LABEL
```
Outside of DAC name, we don't need the other defines, please remove.

martinjaeger replied:
Yes, the `DT_INST_` defines are also generated in devicetree.conf, e.g.:
```
DT_ST_STM32_DAC_40007400_LABEL="DAC_1"
DT_INST_0_ST_STM32_DAC_LABEL="DAC_1"
```
But I just implemented it similar to other drivers like the ADC, which don't rely on `DT_INST_` defines.

So you suggest to completely remove dts_fixups and let the user (and the sample) use the long `DT_INST_0_ST_STM32_DAC_LABEL` instead of `DT_DAC_1_NAME`?

erwango replied:
I'm not very fond of this way of doing, but the only way to go with DT_INST macros are to create a board alias and reference it in sample.
You can check `tests/drivers/i2c` for reference.

carlescufi replied:
Add an alias in the board's `.dts` file and use the macro generated by that alias in the **sample** (`DT_ALIAS_DAC_`)  but use `DT_INST_` macros in the **driver** itself.

galak reviewed (changes requested):
Would like to see if dt defines could use DT_INST_ style and not require dts fixup.

galak commented:
As this introduces a new API this should be raised in the API meeting.

martinjaeger commented:
> As this introduces a new API this should be raised in the API meeting.

I guess it's not gonna make it into v2.2 then, correct?

erwango commented:
> I guess it's not gonna make it into v2.2 then, correct?

Likely.
Good thing would be for you to join the API review meeting. You'll receive information on how to join by mail

tyrantlizard77 reviewed (commented):
system calls LGTM...it would be even better if this came with a test suite.

galak reviewed (changes requested):
We need a test for the API.

carlescufi commented:
@martinjaeger 
Feedback from the API meeting:

- Add a basic API test
- Add an actual hardware test that loops back a DAC pin into an ADC pin

carlescufi commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,161 @@
+ * @retval -EINVAL  If a parameter with an invalid value has been provided.
+ * @retval -ENOTSUP If the requested resolution is not supported.
+ */
+__syscall int dac_write(struct device *dev,
```
I propose this to be `dac_write_value` or `dac_write_single` so that `dac_write` is reserved for sequences.

martinjaeger committed:
tests: drivers: dac: Add write_value API test

Test runs with nucleo_l073rz board.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
tests: drivers: dac: Add loopback test with ADC

The loopback test reads back the DAC output with the ADC and checks
that it is in a narrow range.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented:
All points as discussed in yesterday's API meeting should be fixed now. Also added two tests and validated both on my nucleo board with:
```
./scripts/sanitycheck -T tests/drivers/dac/ --device-testing --device-serial /dev/ttyACM0 -p nucleo_l073rz
```
This nucleo board is not activated in the board .yaml file to be used as a default test (`testing: default: True`). Should I do this? It seems like otherwise the test is only run if `-p nucleo_l073rz` is specified explicitly.

carlescufi commented:
@galak @henrikbrixandersen please re-review

carlescufi commented:
I propose voting on this PR for inclusion in 2.2.

henrikbrixandersen commented on boards/arm/nucleo_l073rz/nucleo_l073rz.dts:
```diff
@@ -38,6 +38,7 @@
 	aliases {
 		led0 = &green_led_2;
 		sw0 = &user_button;
+		dac1 = &dac1;
```
This should be `dac-1 = &dac1` to match the ADC aliases.

henrikbrixandersen commented on drivers/dac/Kconfig:
```diff
@@ -0,0 +1,24 @@
+menuconfig DAC
+	bool "DAC drivers"
+	# All platforms that implement the DAC driver are now required to
+	# provide relevant DTS entries.
```
I think these comments can be removed.

henrikbrixandersen commented on drivers/dac/dac_handlers.c:
```diff
@@ -0,0 +1,33 @@
+	return z_impl_dac_write_value((struct device *)dev,
+		channel, value, resolution);
+}
+#include <syscalls/dac_write_mrsh.c>
```
This needs to be `#include <syscalls/dac_write_value_mrsh.c>`.

henrikbrixandersen commented on drivers/dac/dac_handlers.c:
```diff
@@ -0,0 +1,33 @@
+#include <syscall_handler.h>
+#include <kernel.h>
+
+static inline int z_vrfy_dac_channel_setup(struct device *dev,
```
In order for these handler to work, you need to add an entry for `dac_driver_api` to the `subsystems` array in `scripts/gen_kobject_list.py`.  Otherwise no handlers will be generated and compilation of this file will fail.

henrikbrixandersen commented on samples/drivers/dac/src/main.c:
```diff
@@ -0,0 +1,45 @@
+	while (1) {
+		/* 12-bit DAC, valid digital input from 0 to 4095 */
+		for (int i = 0; i < 4096; i++) {
+			dac_write_value(dac_dev, DAC_1ST_CHANNEL_ID, i, 12);
```
`12` should be `DAC_RESOLUTION`. The interval for the for-loop should match the resolution (here it is hardcoded for 12 bit resolution).

henrikbrixandersen commented on tests/drivers/dac/dac_api/src/main.c:
```diff
@@ -0,0 +1,28 @@
+
+	ztest_test_suite(dac_basic_test,
+			 ztest_user_unit_test(test_dac_write_value),
+			 ztest_unit_test(test_dac_write_value));
```
I believe it is sufficient to run this test in user mode. No need to run it twice.

henrikbrixandersen commented on tests/drivers/dac/dac_api/src/test_dac.c:
```diff
@@ -0,0 +1,77 @@
+		return TC_FAIL;
+	}
+
+	ret = dac_write_value(dac_dev, DAC_1ST_CHANNEL_ID, 2048, 12);
```
`12` needs to be `DAC_RESOLUTION`. It would be ideal if the `2048` was half of the full scale for any resolution (here it is hardcoded).

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/prj.conf:
```diff
@@ -0,0 +1,5 @@
+
+CONFIG_DAC=y
+CONFIG_ADC=y
+CONFIG_TEST_USERSPACE=y
```
I don't believe this test benefits from running under userspace. It should be enough that the basic API test runs in userspace (also, the test fails in userspace since you never grant access to neither the DAC nor the ADC device in `main.c`).

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/main.c:
```diff
@@ -0,0 +1,25 @@
+void test_main(void)
+{
+	ztest_test_suite(dac_loopback_test,
+			 ztest_user_unit_test(test_dac_loopback),
```
Drop userspace test here.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,128 @@
+		return TC_FAIL;
+	}
+
+	ret = dac_write_value(dac_dev, DAC_1ST_CHANNEL_ID, 2048, 12);
```
`12` needs to be `DAC_RESOLUTION`. It would be ideal if 2048 was half of full scale for any resolution (here it is hardcoded).

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,128 @@
+
+	static s16_t m_sample_buffer[1];
+	struct adc_sequence sequence = {
+		.channels    = 1U << ADC_CHANNEL_ID,
```
Use `BIT(ADC_CHANNEL_ID)` here.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,128 @@
+	/* wait to let DAC output settle */
+	k_sleep(10);
+
+	static s16_t m_sample_buffer[1];
```
Not all ADCs use signed 16 bit values. Still thinking of a good way to make this more generic...

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,128 @@
+
+	ret = adc_read(adc_dev, &sequence);
+	zassert_equal(ret, 0, "adc_read() failed with code %d", ret);
+	zassert_within(m_sample_buffer[0], 2048, 32,
```
This code path makes the assumption that the ADC and DAC resolutions are identical. Perhaps we need to make the expected result board dynamic (as in calculated) or hard-coded per board?

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,128 @@
+		.channels    = 1U << ADC_CHANNEL_ID,
+		.buffer      = m_sample_buffer,
+		.buffer_size = sizeof(m_sample_buffer),
+		.resolution  = 12,
```
`12` needs to be `ADC_RESOLUTION`.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,128 @@
+	k_sleep(10);
+
+	static s16_t m_sample_buffer[1];
+	struct adc_sequence sequence = {
```
Make this `static const struct add_sequence ...` to ensure non-specified members are set to zero.

henrikbrixandersen commented on samples/drivers/dac/src/main.c:
```diff
@@ -0,0 +1,45 @@
+
+static const struct dac_channel_cfg dac_ch1_cfg = {
+	.channel_id  = DAC_1ST_CHANNEL_ID,
+	.buffered = false,
```
You need to provide a DAC reference here.

henrikbrixandersen commented on tests/drivers/dac/dac_api/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+common:
+  tags: dac drivers userspace
+tests:
+  drivers.dac:
```
You need to provide `platform_whitelist: nucleo_l073rz`.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+  tags: dac drivers userspace
+tests:
+  drivers.dac:
+    depends_on: dac
```
You need to provide `platform_whitelist: nucleo_l073rz`.

henrikbrixandersen commented:
FYI: I have implemented two (although very similar) DAC drivers based on the proposed API here: https://github.com/henrikbrixandersen/zephyr/commits/mcux_adc_wip

Both are drivers for NXP DACs (The DAC module present in e.g. the K6x series and DAC32 present in e.g. KE1xF series). I have sample/test support too, but awaiting the fixes mentioned above before pushing those commits.

henrikbrixandersen commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,161 @@
+ * See dac_write_value() for argument descriptions.
+ */
+typedef int (*dac_api_write_value)(struct device *dev,
+			    u8_t channel, u32_t value, u8_t resolution);
```
I am not sure if the resolution really belongs in the `write_value()` API call. Perhaps it is better suited in `struct dac_channel_cfg`? Wouldn't it be more logical to configure the resolution once per channel and then just write a value to a channel afterwards?

martinjaeger replied:
Actually thought the same, but wanted to keep it similar to the ADC. As you marked this as resolved already, I guess we keep it as it is for now?

henrikbrixandersen replied:
If I marked it as resolved it wasn’t on purpose.

martinjaeger replied:
Alright, so should I move the resolution setting to channel cfg and break similarity with the ADC? I would be fine with that, but do we want to wait for feedback from others or just go ahead?

henrikbrixandersen replied:
Let's discuss it on the API meeting Tuesday if you are able to join? I see your point in matching the ADC API, although I would prefer the resolution to be set once there as well.

It's not a cardinal point for me. We can take in the API as-in and then evaluate the resolution parameter as it matures.

martinjaeger replied:
Makes sense. I should be able to join tomorrow.

martinjaeger commented:
Thanks @henrikbrixandersen for the detailed review. Updated all requested changes and also made test and samples more generic so that they lead to the same results for 8 to 12-bit DACs.
16-bit DACs will lead to slower frequency of the signal in the sample because k_wait can only go down to 1 msec. Yes, we could use something else here like k_timer. But I think we should keep it simple, as we want to demonstrate the DAC and not the kernel timer features.

carlescufi commented:
@martinjaeger would you mind using an actual email address on your signed-off-by lines? It might be that one day in the future a maintainer might want to contact you.

martinjaeger commented:
> @martinjaeger would you mind using an actual email address on your signed-off-by lines? It might be that one day in the future a maintainer might want to contact you.

I'm a bit concerned about spam bots finding the mail address. And I guess obfuscated email addresses with the at sign written as a word won't be accepted by the CI checks... but you are right, it makes sense to have a usable address in the sign-off. I will change it with next update and hope that the spam bots don't crawl git logs ;)

henrikbrixandersen commented:
> I'm a bit concerned about spam bots finding the mail address. And I guess obfuscated email addresses with the at sign written as a word won't be accepted by the CI checks... but you are right, it makes sense to have a usable address in the sign-off. I will change it with next update and hope that the spam bots don't crawl git logs ;)

Once this is fixed I will approve. I have updated the two NXP DAC drivers to match and adapted the tests for them. I have left out adapting the sample for now.

carlescufi commented:
API meeting: move `u8_t resolution` from the API call the channel configuration structure.

galak commented on drivers/dac/Kconfig.stm32:
```diff
@@ -0,0 +1,35 @@
+
+if DAC_STM32
+
+config DAC_1
```
Instead of doing this we can us DT_INST_.. defines in the driver.

martinjaeger replied:
I'm currently implementing this change, completely removing the `CONFIG_DAC_x` Kconfig settings.
However, I'm not sure how to implement pinmux.c in the board file. It's currently
```
#ifdef CONFIG_DAC_1
	{STM32_PIN_PA4, STM32L0_PINMUX_FUNC_PA4_DAC_OUT1},
#endif /* CONFIG_DAC_1 */
```
Should I use `DT_INST_0_ST_STM32_DAC` again? I think that's confusing because the DAC is named dac1 in the .dts file (aligned with naming in STM32, which starts from 1). I could also use `#ifdef DT_ALIAS_DAC_1_LABEL`, but this doesn't seem very nice either, as it would require to define an alias.

erwango replied:
@martinjaeger, `DT_INST_0_ST_STM32_DAC` should indeed not be used in case of pinmux file. You should keep using `CONFIG_DAC_1` in this case (soon to be replaced by a new `DT_NODELABEL_FOO` macro.
This is an on going change (https://github.com/zephyrproject-rtos/zephyr/issues/19285 provides an overview). Sorry for the confusion it creates in your  porting work.

martinjaeger replied:
Alright, so I'll revert the changes again and keep the Kconfig settings.

galak replied:
So are there driver reverts that are needed?

martinjaeger replied:
No, should be fine as it is in this PR.

galak commented on drivers/dac/dac_stm32.c:
```diff
@@ -0,0 +1,180 @@
+
+/* DT instance numbering starts from 0, STM DAC from 1 */
+
+#ifdef CONFIG_DAC_1
```
Change these from `#ifdef CONFIG_DAC_1` to `#ifdef DT_INST_0_ST_STM32_DAC`

martinjaeger replied:
Done.

galak commented on tests/drivers/dac/dac_api/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+  tags: dac drivers userspace
+tests:
+  drivers.dac:
+    platform_whitelist: nucleo_l073rz
```
Can we make this `dt_compat_enabled_with_alias` look at samples/basic/threads/sample.yaml for example.

henrikbrixandersen replied:
That won't be sufficient. The source file contains board-specific code.

martinjaeger replied:
So I assume we can keep this as it is?

galak reviewed (dismissed):
Some minor changes to the driver to remove Kconfig instance symbols.

henrikbrixandersen commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,163 @@
+	/**
+	 * Enable buffered output if available in underlying hardware.
+	 */
+	bool buffered;
```
It’s a quite common feature. I would like to keep it in the the API.

mnkp replied:
Could you give a few examples? Just to get a better feeling on how the feature is implemented / documented by different vendors.

henrikbrixandersen replied:
@mnkp Are you asking for examples of devices implementing a buffered DAC output? They are quite common. See the NXP Kinetis DAC32 for instance.

martinjaeger replied:
Also Atmel SAM microcontrollers have a buffered output, see [here](http://ww1.microchip.com/downloads/en/AppNotes/atmel-42110-sam0-digital-to-analog-(dac)-driver_application%20note_at03244.pdf), Section 7.4.7. Configuration and Initialization (Channel). They specifically mention that the buffer should be switched off for power saving if not needed, so it has to be configurable at runtime.

mnkp replied:
Thanks, yes I was asking for existing examples of how the "buffered output" feature is implemented by other vendors.

The NXP [DAC32: Digital-to-Analog Converter](https://mcuxpresso.nxp.com/api_doc/dev/1008/group__dac32.html) is a DAC API from NXP and when they talk about DAC buffer they mean a hardware buffer storing the samples. E.g. in case of K64 family it's a 16-word data buffer.

In the example given by @martinjaeger regarding SAM0 family DAC the buffered output refers to the possibility of enabling or disabling analog output buffer. E.g. with DAC buffer disabled the VOUT pin will not be driven, though the Vout voltage from the DAC can still be routed to an internal module.

As implemented in this PR the buffered option enables "higher drive current capability" of the analog output. If the option is disabled the analog output is still driven however with "lower drive current capability".

Every of the three examples talks about a different feature.

henrikbrixandersen replied:
@mnkp You are talking about the NXP DAC (not DAC32) present in the K64 family, which - as you say - does not implement a hardware buffer. I am talking about the NXP DAC32 present in e.g. the KE1xF series, which does implement a hardware buffer (and no, I can tell the difference between a hardware buffer and a data buffer).

Please see the following block diagram for the NXP DAC32 from the KE1xF reference manual which clearly shows an "AMP buffer":

<img width="547" alt="Screenshot 2020-03-30 at 07 10 58" src="https://user-images.githubusercontent.com/1076226/77877456-1f884680-7256-11ea-80c4-20a81e4caea9.png">

henrikbrixandersen replied:
On a side note, I have already implemented DAC drivers for the NXP DAC and DAC32 here:  https://github.com/henrikbrixandersen/zephyr/commits/mcux_dac_wip

martinjaeger replied:
> Every of the three examples talks about a different feature.

I think it's the same thing for the mentioned Atmel SAM devices and the STM32. The SAM can only drive external pins with enabled buffer, but it's still a useful configuration option to switch it on or off. In case of STM32 you can drive an external pin even without buffer, but you get a much higher impedance output. So you can select if you need the output buffer depending on the application.

henrikbrixandersen replied:
Another option could be to implement the power management API for the DACs where the output can be turned off to save power - and convert the hardware buffer enable/disable to a devicetree setting.

mnkp replied:
Thanks for the clarification. The behavior of NXP KE1xF looks similar to SAM0 family DAC in regards to a possibility of enabling/disabling the output buffer/amplifier. I checked a few more datasheets, it does look like a fairly popular feature.

Even if there are differences between vendors, like STM32 which allows to disable the buffer but still routes the analog signal to the output pad, it is a pretty consistent behavior.

However, the parameter description is a bit misleading at the moment
```
 * @param buffered Enable buffered output if available in underlying hardware.
```
We should be more explicit that we are talking about the analog output buffer. Also, every DAC will have some kind of analog buffer/amp that will drive the signal on the pad so user should always enable it if they intend to use it. E.g. in case of NXP K64 family it's not possible to disable the output buffer, as such the driver should return -ENOTSUP if the buffered parameter is set to false. If we believe that's not the intended behavior, i.e. if the driver doesn't support the feature it should be ignored this should also be documented.

For the usability it may be better to invert the logic and give user the option to disable the buffer rather than require him/her to explicitly enable it. e.g.
```
 * @param out_buf_disable Disable analog output buffer.
```

martinjaeger commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,163 @@
+ * @return a positive value is the reference voltage value. Returns
+ * zero if reference voltage information is not available.
+ */
+static inline u16_t dac_ref_internal(struct device *dev)
```
This is named the same as for the ADC. I think it should be consistent.

martinjaeger commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,163 @@
+
+	/**
+	 * Enable buffered output if available in underlying hardware.
+	 */
```
Interesting. Thanks for the comment. That also explains why the ADC comments are not rendered properly, which I took as a basis.
It must be a Zephyr-specific Doxygen option, as that way of commenting would be rendered correctly in another project where I use Doxygen aswell.

Will update it.

martinjaeger commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,163 @@
+ * See dac_write_value() for argument descriptions.
+ */
+typedef int (*dac_api_write_value)(struct device *dev,
+				    u8_t channel, u32_t value);
```
While I agree with you, this is actually not used by "all other Zephyr APIs" consistently. The ADC driver (which I took as a main reference) doesn't have it. And also more recently added drivers like the LoRa driver expose the internal API functions to Doxygen.

So who is the official target audience for the final documentation? Only users of the final drivers or also developers who want to port the driver to other targets for example?

mnkp replied:
This is official API documentation targeting end users. Thanks for fixing.

carlescufi commented:
@mnkp and @galak are you OK with merging this?

erwango commented:
@martinjaeger, can you rebase ?

martinjaeger commented:
> @martinjaeger, can you rebase ?

Here we go.

The failure in Shipable seems to be unrelated to this PR.

erwango commented:
Shippable error:
` Sanitycheck / qemu_x86_nommu:tests/kernel/timer/timer_api/kernel.timer.tickless / tests/kernel/timer/timer_api/kernel.timer.tickless`

@galak, @mnkp

carlescufi commented:
@martinjaeger could you please rebase and fix the `k_sleep()` issues? you need to use `k_msleep()` or `K_MSEC()`

carlescufi commented:
Also, @martinjaeger could you please add entries here:

- New .rst file in https://github.com/zephyrproject-rtos/zephyr/tree/master/doc/reference/peripherals
- Entry in the table https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/overview.rst

martinjaeger commented:
@carlescufi Sure, will rebase and fix the issues you mentioned. Should I also change to new DT macros?

carlescufi commented:
> @carlescufi Sure, will rebase and fix the issues you mentioned. Should I also change to new DT macros?

Well that would be great. Thanks @martinjaeger for all the effort. I am hoping to get an ACK for this today.

pabigot commented on include/drivers/dac.h:
```diff
@@ -0,0 +1,165 @@
+__subsystem struct dac_driver_api {
+	dac_api_channel_setup channel_setup;
+	dac_api_write_value   write_value;
+	u16_t ref_internal;	/* mV */
```
Repeating the comment that didn't get saved: This doesn't handle all cases, as described in https://github.com/zephyrproject-rtos/zephyr/pull/24292#issuecomment-612605361 so it might be worth not following ADC in this case.

carlescufi commented:
API meeting:

- Remove the buffered output variable from the configuration (for now, to be reworked with a combo of DT + config())
- Remove the `ref_internal` variable, to be reworked as well with DT.

henrikbrixandersen commented on doc/reference/peripherals/dac.rst:
```diff
@@ -0,0 +1,14 @@
+
+Overview
+********
+
```
Agreed. It does not have to be much, initially. How about something like this:

```
The DAC API provides access to Digital-to-Analog Converter (DAC) devices.

Configuration Options
*********************

Related configuration options:

* :option:`CONFIG_DAC`
```

henrikbrixandersen commented on samples/drivers/dac/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
This needs to be updated to `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`.

henrikbrixandersen commented on samples/drivers/dac/src/main.c:
```diff
@@ -0,0 +1,56 @@
+#include <sys/printk.h>
+#include <drivers/dac.h>
+
+#define DAC_DEVICE_NAME		DT_ALIAS_DAC_1_LABEL
```
This should be updated to `#define DAC_DEVICE_NAME DT_LABEL(DT_ALIAS(dac_1))`

henrikbrixandersen commented on tests/drivers/dac/dac_api/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Use `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})` here as well.

henrikbrixandersen commented on tests/drivers/dac/dac_api/src/test_dac.c:
```diff
@@ -0,0 +1,77 @@
+
+#if defined(CONFIG_BOARD_NUCLEO_L073RZ)
+
+#define DAC_DEVICE_NAME		DT_ALIAS_DAC_1_LABEL
```
Use `DT_LABEL(DT_ALIAS(dac_1))` here as well.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Use `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})` here as well.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,132 @@
+
+/* DAC output on PA4, ADC input read from PA1 */
+
+#define DAC_DEVICE_NAME		DT_ALIAS_DAC_1_LABEL
```
Convert to new DT API.

henrikbrixandersen commented on tests/drivers/dac/dac_loopback/src/test_dac.c:
```diff
@@ -0,0 +1,132 @@
+#define DAC_1ST_CHANNEL_ID	1
+#define DAC_RESOLUTION		12
+
+#define ADC_DEVICE_NAME		DT_ADC_1_NAME
```
Convert to new DT API.

erwango commented on boards/arm/nucleo_l073rz/pinmux.c:
```diff
@@ -34,6 +34,9 @@ static const struct pin_config pinconf[] = {
 #ifdef CONFIG_ADC_1
 	{STM32_PIN_PA0, STM32L0_PINMUX_FUNC_PA0_ADC_IN0},
 #endif /* CONFIG_ADC_1 */
+#ifdef CONFIG_DAC_1
```
I think you can use the following:
```
#if DT_HAS_NODE(DT_NODELABEL(dac1))
```

erwango commented on drivers/dac/Kconfig.stm32:
```diff
@@ -0,0 +1,35 @@
+
+if DAC_STM32
+
+config DAC_1
+	bool "STM32 DAC 1"
+	help
+	  Enable DAC 1. Each STM32 DAC has 2 channels.
+
+config DAC_2
+	bool "STM32 DAC 2"
+	help
+	  Enable DAC 2. Each STM32 DAC has 2 channels.
+
+config DAC_3
+	bool "STM32 DAC 3"
+	help
+	  Enable DAC 3. Each STM32 DAC has 2 channels.
+
+config DAC_4
+	bool "STM32 DAC 4"
+	help
+	  Enable DAC 4. Each STM32 DAC has 2 channels.
+
```
These symbols should not be needed if using suggested tip in board pinmux files

erwango commented on boards/arm/nucleo_l073rz/Kconfig.defconfig:
```diff
@@ -30,4 +30,11 @@ config ADC_1
 	default y
 	depends on ADC
 
+if DAC
+
+config DAC_1
+	default y
+
+endif # DAC
```
Should not be needed, cf below comment

erwango commented on drivers/dac/Kconfig.stm32:
```diff
@@ -0,0 +1,35 @@
+
+menuconfig DAC_STM32
+	bool "STM32 DAC driver"
+	depends on SOC_FAMILY_STM32
```
```
	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DAC))
```
with 
```
DT_COMPAT_ST_STM32_DAC := st,stm32-dac
```

erwango commented on soc/arm/st_stm32/stm32l0/Kconfig.defconfig.series:
```diff
@@ -20,4 +20,11 @@ config DMA_STM32_V2
 	default y
 	depends on DMA_STM32
 
+if DAC
+
+config DAC_STM32
+	default y
+
+endif # DAC
```
Not needed if Kconfig configured thanks to `dt_compat_enabled`

martinjaeger commented:
Thanks @erwango and @henrikbrixandersen for your additional review. It's sometimes a bit difficult to keep track of all changes happening in the background, e.g. new DT macros and those changes in CMakeLists.txt. But I hope it's now all up-to-date and can finally be merged.

With the DT and Kconfig updates we could remove >100 lines of redundant code. Nice.

Changes decided in API meeting are also included @carlescufi.

erwango commented on boards/arm/nucleo_l073rz/nucleo_l073rz.dts:
```diff
@@ -38,6 +38,7 @@
 	aliases {
 		led0 = &green_led_2;
 		sw0 = &user_button;
+		dac-1 = &dac1;
```
I wonder if we shouldn't use uniform alias names: dac0.

erwango replied:
We miss the adc alias

henrikbrixandersen reviewed (approved):
Looks good!

erwango commented on drivers/dac/dac_stm32.c:
```diff
@@ -0,0 +1,165 @@
+	struct dac_stm32_data *data = dev->driver_data;
+	const struct dac_stm32_cfg *cfg = dev->config->config_info;
+
+	if (channel_cfg->channel_id - STM32_FIRST_CHANNEL >=
+			data->channel_count ||
+			channel_cfg->channel_id < STM32_FIRST_CHANNEL) {
```
Please add parenthesis on OR operands

erwango commented on drivers/dac/dac_stm32.c:
```diff
@@ -0,0 +1,165 @@
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution == 8 || channel_cfg->resolution == 12) {
```
ditto

erwango reviewed (changes requested):
Few more minor comments + missing alias to get loopback test compiling.


Also, I think you can squash "drivers: dac: Convert to new DT_INST macros" into "drivers: dac: Add support for STM32L0 series".

I don't have nucleo_l073rz, so I've tried to test on nucleo_l053r8 which share the same pinmux. Api test was working but loopback test failed (bridging Arduino A0/A2):
```
    Assertion failed at ../../src/test_dac.c:121: test_task_loopback: (m_sample)
Value 615 read from ADC does not match expected range.     
```
Any clue why this is not working ?

martinjaeger commented:
> Also, I think you can squash "drivers: dac: Convert to new DT_INST macros" into "drivers: dac: Add support for STM32L0 series".

Done

> I don't have nucleo_l073rz, so I've tried to test on nucleo_l053r8 which share the same pinmux. Api test was working but loopback test failed (bridging Arduino A0/A2):
> 
> ```
>     Assertion failed at ../../src/test_dac.c:121: test_task_loopback: (m_sample)
> Value 615 read from ADC does not match expected range.     
> ```
> 
> Any clue why this is not working ?

For the Nucleo L073 you need to bridge A1 and A2.

But you were right regarding the compile error in the loopback test, of course. Fixed that and renamed the aliases to adc0 and dac0 as suggested. Both tests are compiling and passing now.

martinjaeger commented on drivers/dac/Kconfig.stm32:
```diff
@@ -0,0 +1,15 @@
+
+config DAC_STM32
+	bool "STM32 DAC driver"
+	depends on SOC_FAMILY_STM32
```
Oops, forgot this one. One more rebase...

erwango reviewed (approved):
Approved

> For the Nucleo L073 you need to bridge A1 and A2.

Bridging A1/A2 work on nucleo_l053r8 works on my side as well.
But my eyes still tell me this should not work as ADC1 IN0 is set to PA0 which is A0 according to https://www.st.com/resource/en/user_manual/dm00105823-stm32-nucleo-64-boards-mb1136-stmicroelectronics.pdf.
Unless this doc is wrong and then arduino_r3_connector.dtsi should be fixed for these 2 boards.

martinjaeger commented:
> Bridging A1/A2 work on nucleo_l053r8 works on my side as well.
> But my eyes still tell me this should not work as ADC1 IN0 is set to PA0 which is A0 according to https://www.st.com/resource/en/user_manual/dm00105823-stm32-nucleo-64-boards-mb1136-stmicroelectronics.pdf.
> Unless this doc is wrong and then arduino_r3_connector.dtsi should be fixed for these 2 boards.

We are reading ADC1_IN1 at PA1 (see tests/drivers/dac/dac_loopback/src/test_dac.c#L41):
```
#define ADC_DEVICE_NAME		DT_LABEL(DT_ALIAS(adc1))
#define ADC_CHANNEL_ID		1
```
ADC channels also start counting at 0 in the driver.

martinjaeger commented:
Can someone please explain what's the issue with Shippable? From the build log I don't understand what's wrong exactly or if it is at all caused by the changes introduced in this PR.

erwango commented:
> Can someone please explain what's the issue with Shippable? From the build log I don't understand what's wrong exactly or if it is at all caused by the changes introduced in this PR.

I don't think it is linked. QEMU testing happens to be unstable. You can try to close/reopen the PR to launch a new shippable run and see if it gets fixed.

martinjaeger commented on tests/drivers/dac/dac_loopback/boards/nucleo_l073rz.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_ADC_1=y
```
Ok, removed it. Gives us another chance to pass Shippable :)

erwango replied:
and it passed \o/

carlescufi commented:
@martinjaeger thank you very much for this contribution and for your patience driving it through.
